embassy-stm32/src/dma/mod.rs

```diff
@@ -29,27 +30,91 @@ pub trait Channel: sealed::Channel {
     where
         Self: 'a;
 
-    fn read<'a>(
+    type CompletionFuture<'a>: Future<Output = ()> + 'a
+    where
+        Self: 'a;
+
+    fn read_u8<'a>(
         &'a mut self,
         request: Request,
-        src: *mut u8,
+        reg_addr: *mut u32,
         buf: &'a mut [u8],
     ) -> Self::ReadFuture<'a>;
 
-    fn write<'a>(
+    fn read_u16<'a>(
+        &'a mut self,
+        request: Request,
+        reg_addr: *mut u32,
+        buf: &'a mut [u16],
+    ) -> Self::ReadFuture<'a>;
+
+    fn read_u32<'a>(
+        &'a mut self,
+        request: Request,
+        reg_addr: *mut u32,
+        buf: &'a mut [u32],
+    ) -> Self::ReadFuture<'a>;
+
+    fn write_u8<'a>(
         &'a mut self,
         request: Request,
         buf: &'a [u8],
-        dst: *mut u8,
+        reg_addr: *mut u32,
+    ) -> Self::WriteFuture<'a>;
+
+    fn write_u16<'a>(
+        &'a mut self,
+        request: Request,
+        buf: &'a [u16],
+        reg_addr: *mut u32,
+    ) -> Self::WriteFuture<'a>;
+
+    fn write_u32<'a>(
+        &'a mut self,
+        request: Request,
+        buf: &'a [u32],
+        reg_addr: *mut u32,
     ) -> Self::WriteFuture<'a>;
 
     fn write_x<'a>(
         &'a mut self,
         request: Request,
         word: &u8,
         num: usize,
-        dst: *mut u8,
+        reg_addr: *mut u32,
     ) -> Self::WriteFuture<'a>;
+
+    /// Starts this channel for writing a stream of bytes.
+    fn start_write_u8<'a>(&'a mut self, request: Request, buf: &'a [u8], reg_addr: *mut u32);
+
+    /// Starts this channel for writing a stream of half-words.
+    fn start_write_u16<'a>(&'a mut self, request: Request, buf: &'a [u16], reg_addr: *mut u32);
```

Comment:
Maybe this could be generic over a `Word` trait impl'd for `u8, u16, u32`, so that there's just `start_read<W>, start_write<W>`. This would remove the need for the ugly macros in the impl side.